### PR TITLE
fix(build): switch to esbuild and tsc from pika

### PR DESCRIPTION
**Pull request by kfcampbell**
Replicating [octokit/request.js#584](https://github.com/octokit/request.js/pull/584) to another repo. 

---

Pika has been deprecated for a while now, and the project has now been archived, and because of that the released npm package is missing most of the files generated by the build step.

Uses esbuild to transpile the TS source code into an ESM source, NodeJs bundle, and a browser bundle

Uses the TypeScript compiler to generate the types.

### Before the change?

* Uses pika for the build system
* The released npm package is missing most of the files generated by the build step. `dist-node`, `dist-types`, `dist-web`... even though they are generated correctly.

### After the change?

* Uses esbuild to generate the JS bundles
* Uses the TypeScript compiler to generate the type definitions
* The correct file matching patterns are used so we publish all the necessary files again.

----

### Pull request checklist
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Added the appropriate label for the given change

### Does this introduce a breaking change?

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/main/community/breaking_changes.md) to help!

- [ ] Yes (Please add the `Type: Breaking change` label)
- [x] No

If `Yes`, what's the impact:

* N/A

### Pull request type

Please add the corresponding label for change this PR introduces:
- Bugfix: `Type: Bug`
- Dependencies/code cleanup: `Type: Maintenance`

----

### CODE REVIEW

**Review by nickfloyd — APPROVED** [review]
🎉

**Comment by wolfy1339** [comment]
Don't forget to mark all type imports and exports

**Comment by gr2m** [comment]
> Don't forget to mark all type imports and exports

I've seen this causing a few times. Do you know if we can somehow add a CI check to make sure we don't miss these in future?

**Comment by kfcampbell** [comment]
@wolfy1339 I believe I've done this correctly now; please take another look!

@gr2m we could probably use something like [this eslint rule](https://typescript-eslint.io/rules/consistent-type-imports/). Without taking an extra dependency, perhaps it'd be better to use [`--verbatimModuleSyntax`](https://www.typescriptlang.org/tsconfig#verbatimModuleSyntax) when compiling. Thoughts?

**Comment by gr2m** [comment]
> @gr2m we could probably use something like [this eslint rule](https://typescript-eslint.io/rules/consistent-type-imports/). Without taking an extra dependency, perhaps it'd be better to use [`--verbatimModuleSyntax`](https://www.typescriptlang.org/tsconfig#verbatimModuleSyntax) when compiling. Thoughts?

yeah I would like to avoid adding eslint

`--verbatimModuleSyntax` looks great, I didn't know about this flag! It seems like the only gotcha about it is for modules with side effect, which we have been careful to avoid. I'd say let's add it to all the things

**Comment by wolfy1339** [comment]
I agree that the typescript build option looks good.

However I'm not sure if it will have an impact since esbuild doesn't use typescript to compile.

**Comment by kfcampbell** [comment]
:+1: on adding the build flag; I've done so in the latest commit. When the type keyword is removed from imports ([here](https://github.com/octokit/auth-token.js/blob/switch-to-esbuild/src/hook.ts#L1) for example), the `tsc` call in our `npm run build` command throws the following errors:

```
 sh$ npm run build

> @octokit/auth-token@0.0.0-development build /home/kfcampbell/github/dev/auth-token.js
> node scripts/build.mjs && tsc -p tsconfig.json --verbatimModuleSyntax

src/hook.ts:2:3 - error TS1484: 'AnyResponse' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.

2   AnyResponse,
    ~~~~~~~~~~~

src/hook.ts:3:3 - error TS1484: 'EndpointDefaults' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.

3   EndpointDefaults,
    ~~~~~~~~~~~~~~~~

src/hook.ts:4:3 - error TS1484: 'EndpointOptions' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.

4   EndpointOptions,
    ~~~~~~~~~~~~~~~

src/hook.ts:5:3 - error TS1484: 'RequestInterface' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.

5   RequestInterface,
    ~~~~~~~~~~~~~~~~

src/hook.ts:6:3 - error TS1484: 'RequestParameters' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.

6   RequestParameters,
    ~~~~~~~~~~~~~~~~~

src/hook.ts:7:3 - error TS1484: 'Route' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.

7   Route,
    ~~~~~

src/hook.ts:8:3 - error TS1484: 'Token' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.

8   Token,
    ~~~~~


Found 7 errors in the same file, starting at: src/hook.ts:2

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! @octokit/auth-token@0.0.0-development build: `node scripts/build.mjs && tsc -p tsconfig.json --verbatimModuleSyntax`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the @octokit/auth-token@0.0.0-development build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/kfcampbell/.npm/_logs/2023-06-05T22_15_33_319Z-debug.log
```

**Comment by gr2m** [comment]
That's great! I'd add it as an option to https://github.com/octokit/tsconfig and maybe release it as a breaking change so that the different packages don't suddenly start breaking. We can then address all type import errors like show in your comment as we update to the new breaking version of `@octokit/tsconfig`

**Comment by wolfy1339** [comment]
While we're at it, we should move the options that are common across the repos to `@octokit/tsconfig`

**Comment by wolfy1339** [comment]
> :+1: on adding the build flag; I've done so in the latest commit. When the type keyword is removed from imports ([here](https://github.com/octokit/auth-token.js/blob/switch-to-esbuild/src/hook.ts#L1) for example), the `tsc` call in our `npm run build` command throws the following errors:
> 
> ```
>  sh$ npm run build
> 
> > @octokit/auth-token@0.0.0-development build /home/kfcampbell/github/dev/auth-token.js
> > node scripts/build.mjs && tsc -p tsconfig.json --verbatimModuleSyntax
> 
> src/hook.ts:2:3 - error TS1484: 'AnyResponse' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.
> 
> 2   AnyResponse,
>     ~~~~~~~~~~~
> 
> src/hook.ts:3:3 - error TS1484: 'EndpointDefaults' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.
> 
> 3   EndpointDefaults,
>     ~~~~~~~~~~~~~~~~
> 
> src/hook.ts:4:3 - error TS1484: 'EndpointOptions' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.
> 
> 4   EndpointOptions,
>     ~~~~~~~~~~~~~~~
> 
> src/hook.ts:5:3 - error TS1484: 'RequestInterface' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.
> 
> 5   RequestInterface,
>     ~~~~~~~~~~~~~~~~
> 
> src/hook.ts:6:3 - error TS1484: 'RequestParameters' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.
> 
> 6   RequestParameters,
>     ~~~~~~~~~~~~~~~~~
> 
> src/hook.ts:7:3 - error TS1484: 'Route' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.
> 
> 7   Route,
>     ~~~~~
> 
> src/hook.ts:8:3 - error TS1484: 'Token' is a type and must be imported using a type-only import when 'verbatimModuleSyntax' is enabled.
> 
> 8   Token,
>     ~~~~~
> 
> 
> Found 7 errors in the same file, starting at: src/hook.ts:2
> 
> npm ERR! code ELIFECYCLE
> npm ERR! errno 2
> npm ERR! @octokit/auth-token@0.0.0-development build: `node scripts/build.mjs && tsc -p tsconfig.json --verbatimModuleSyntax`
> npm ERR! Exit status 2
> npm ERR!
> npm ERR! Failed at the @octokit/auth-token@0.0.0-development build script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /home/kfcampbell/.npm/_logs/2023-06-05T22_15_33_319Z-debug.log
> ```

That doesn't catch the ones which aren't types but are used as types, like when using a class for the type of a parameter in a function

File: package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
   "version": "0.0.0-development",
   "description": "GitHub API token authentication for browsers and Node.js",
   "scripts": {
-    "build": "pika-pack build",
+    "build": "node scripts/build.mjs && tsc -p tsconfig.json --verbatimModuleSyntax",
```
Comment (wolfy1339) [inline package.json:9]: Can you move this into the `tsconfig.json`

**Comment by kfcampbell** [comment]
> I'd add it as an option to https://github.com/octokit/tsconfig and maybe release it as a breaking change so that the different packages don't suddenly start breaking. We can then address all type import errors like show in your comment as we update to the new breaking version of @octokit/tsconfig

@gr2m that seems reasonable! I've created [octokit/tsconfig#15](https://github.com/octokit/tsconfig/pull/15) to do so. Once that PR is merged and released, I can update all the in-flight pika removal PRs to bump octokit/tsconfig versions as well. 

> That doesn't catch the ones which aren't types but are used as types, like when using a class for the type of a parameter in a function

@wolfy1339 if you can find a good way to do this, please let me know and I can help roll that out as well! Thanks for all the thoughts so far.

**Comment by kfcampbell** [comment]
> While we're at it, we should move the options that are common across the repos to @octokit/tsconfig

Some other options that may be reasonable to move to octokit/tsconfig are:

```json
  "compilerOptions": {
    "esModuleInterop": true,
    "declaration": true,
    "emitDeclarationOnly": true,
    "sourceMap": true,
  },
  "include": ["src/**/*"]
}
```

Do either of you have thoughts on these or others before I add them to octokit/tsconfig?

**Comment by kfcampbell** [comment]
Note that as of [9a3587f](https://github.com/octokit/auth-token.js/pull/345/commits/9a3587f7e65d73051f78472e0b7a5311c3cc64c9), tests are failing in Node 16 with errors like:

```
    test/index.test.ts:1:10 - error TS1286: ESM syntax is not allowed in a CommonJS module when 'verbatimModuleSyntax' is enabled.

    1 import { request } from "@octokit/request";
```

I've resolved these instances by using the CommonJS syntax `const { request } = require("@octokit/request");`

However, we're also getting an issue in index.ts: 

```
 FAIL  test/index.test.ts
  ● Test suite failed to run

    src/index.ts:11:1 - error TS1287: A top-level 'export' modifier cannot be used on value declarations in a CommonJS module when 'verbatimModuleSyntax' is enabled.

    11 export const createTokenAuth: StrategyInterface = function createTokenAuth(
       ~~~~~~
```

We could use an intermediary object in exports, though that would be a breaking change for consumers. Switching to ESM would be an option, though that's a whole other effort we've been postponing for some time. Are there other options I'm missing here to export this correctly?

**Comment by wolfy1339** [comment]
I think we can add a tsconfig just for tests that disables that option

**Comment by kfcampbell** [comment]
I'm picturing changing `package.json` to read something like:

```json
...
  "scripts": {
    "build": "node scripts/build.mjs && tsc -p tsconfig.json",
    "test": "node scripts/build.mjs && tsc -p tsconfig.test.json && jest --coverage",
...
```

I don't love the idea of splitting the files but it's probably better than avoiding the compiler option entirely, especially with how easy it is (for me) to forget about type exports. Is that also what you're picturing?

**Comment by wolfy1339** [comment]
> I'm picturing changing `package.json` to read something like:
> 
> ```json
> ...
>   "scripts": {
>     "build": "node scripts/build.mjs && tsc -p tsconfig.json",
>     "test": "node scripts/build.mjs && tsc -p tsconfig.test.json && jest --coverage",
> ...
> ```
> 
> I don't love the idea of splitting the files but it's probably better than avoiding the compiler option entirely, especially with how easy it is (for me) to forget about type exports. Is that also what you're picturing? 

You don't need to change the scripts at all.
The only change that could be necessary is to the jest config to point it to the correct tsconfig

**Comment by kfcampbell** [comment]
@wolfy1339 thanks for the help! It's now done. I've also bumped the major version of octokit/tsconfig and removed the explicit specification of `verbatimModuleSyntax`. Please take another look!

File: tsconfig.test.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "@octokit/tsconfig",
+  "compilerOptions": {
+    "esModuleInterop": true,
+    "declaration": true,
+    "outDir": "pkg/dist-types",
+    "emitDeclarationOnly": true,
+    "sourceMap": true,
+    "verbatimModuleSyntax": false
```
Comment (wolfy1339) [inline tsconfig.test.json:9]:
```suggestion
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "emitDeclerarionOnly": false
    "noEmit": true,
    "verbatimModuleSyntax": false
```

File: tsconfig.test.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  },
+  "include": ["src/**/*"]
```
Comment (wolfy1339) [inline tsconfig.test.json:11]:
```suggestion
  "include": ["test/**/*"]
```

Comment (wolfy1339) [reply tsconfig.test.json:11]: This one I'm less sure about.

**Review by wolfy1339 — COMMENTED** [review]
Would it be wise to move it into the test directory?

As done in `@octokit/webhooks`:
https://github.com/octokit/webhooks.js/blob/bc0ab962cb6218b6714c594a4b1c16ec8acbb707/tsconfig.json

File: tsconfig.test.json (hunk 3 above)
Comment (kfcampbell) [reply tsconfig.test.json]: I don't think we should take this: then the tsconfig would only apply to the single test file, not the source code. Perhaps `include` should be something like `**/*` if we want to include both the test file(s) and the source code in the test tsconfig.

**Comment by kfcampbell** [comment]
> Would it be wise to move it into the test directory?

Good call! Done.

**Review by wolfy1339 — APPROVED** [review]

**Comment by kfcampbell** [comment]
Once merged, any reservations to me rolling out the pattern introduced on this PR to the other octokit modules using pika?

**Pull request merged**